src/decoder/cuda-lattice-decoder.cu

```diff
@@ -0,0 +1,1652 @@
+    int32 size = GetSize(arcs_bpr_fr_sidx_d, frame);
+    for (; tid < size; tid += gridDim.x * blockDim.x) {
+      LatLink* link = GetActiveArc(frame, tid);
+      Token* next_tok = GetActiveToken(link->p1, true, frame);
+      Token* tok = GetActiveToken(link->p2, true, frame);
+      BaseFloat link_extra_cost = next_tok->extra_cost +
+                                  ((tok->cost_ + link->acoustic_cost + link->graph_cost)
+                                   - next_tok->cost_);
+      if (!isnan(link_extra_cost) && link_extra_cost <= lattice_beam) {
+        //not pruned out
+        if (merge) {
+          AddArc(link);
+          //link->acoustic_cost=CUDART_NAN_F;
+          //don't need to delete it in original lattice
+        }
+      }
+    }
+    __grid_sync_nv_internal(barrier_);
+  }
+
+  /*
+  { // we do not prune lattice node
+    //update tok
+    int32 tid=threadIdx.x+blockIdx.x*blockDim.x;
+    int32 size=GetSize(toks_bpr_fr_sidx_d,frame);
+    for (;tid<size;tid+=gridDim.x*blockDim.x) {
+      Token* tok=GetActiveToken(frame-1,tid);
+      if (tok->extra_cost==FLT_MAX)
+        tok->tot_cost=CUDART_NAN_F; //prune
+    }
+  } 
+  */
+
+  //get size
+  if (merge && rank0) {
+    int& size_arc_of_frame = arcs_apr_fr_size_d[frame];
+    size_arc_of_frame = *arcs_apr_used_d - prev_cidx;
+    if (verbose > 3) CUDA_PRINTF("PR %i %i %i\n", frame,
+            GetSize(arcs_bpr_fr_sidx_d, frame), size_arc_of_frame);
+  }
+  //__grid_sync_nv_internal(barrier_);
+}
+
+// copy accumulated arcs after lattice pruning till the given frame
+// after obtaining the copy size, copy the buffer asynchronously
+void LatticePruner::CopyArcsToHost(int32 frame, cudaStream_t st) {
+  int32 sz;
+  cudaMemcpy(arcs_apr_used_h, arcs_apr_used_d,
+             sizeof(int32), cudaMemcpyDeviceToHost);
+  // TODO: optimize out above overhead
+  // one possibility is we can copy static length
+  // by assuming ESTIMATED_PRUNE_RATIO parts are remained
+  //sz=sizeof(LatLink)*(arcs_buf_before_pr_size*ESTIMATED_PRUNE_RATIO); 
+
+  sz = sizeof(LatLink) * (*arcs_apr_used_h); // use exact count
+  cudaMemcpyAsync(arcs_apr_h, arcs_apr_d,
+                  sz, cudaMemcpyDeviceToHost, st);
+  sz = sizeof(int32) * (frame + 1) * (1);
+  cudaMemcpyAsync(arcs_apr_fr_size_h, arcs_apr_fr_size_d,
+                  sz, cudaMemcpyDeviceToHost, st);
+  // clear arcs_apr_used_d in GPU during next call of pruning 
+}
+
+// copy accumulated toks till the given frame
+// after obtaining the copy size, copy the buffer asynchronously
+void LatticePruner::CopyToksToHost(int32 frame, cudaStream_t st) {
+  int32 sz;
+  //include frame 0 count and the total count in the last element
+  sz = sizeof(int32) * (frame + 1 + 1) * (1); 
+  cudaMemcpy(toks_bpr_fr_sidx_h, toks_bpr_fr_sidx_d,
+             sz, cudaMemcpyDeviceToHost);
+  sz = sizeof(Token) * (toks_bpr_fr_sidx_h[frame + 1]);
+  assert(sz); // assume we have obtain the total count 
```

Comment:
you should probably use KALDI_ASSERT instead (and in other places too, unless it's very low level code)

Reply:
it fails to compile in device code